File: Lista de Exercício/Java/Animal.java
package ListaDeExercicio.Java;

public class Animal {

    public static void main(String[] args) {
        Animal cachorro = new Cachorro();
        Animal gato = new Gato();
        Animal vaca = new Vaca();

        Animal[] animais = {cachorro, gato, vaca};

        fazerBarulho(animais);
    }

    // Método que faz os animais fazerem barulho
    public static void fazerBarulho(Animal[] animais) {
        for (Animal animal : animais) {
            System.out.println("O " + animal.getTipo() + " faz: " + animal.som());
        }
    }

    public String som() {
        return "";
    }

    // Retorna o tipo do animal baseado no nome da classe
    public String getTipo() {
        return this.getClass().getSimpleName();
    }
}

class Cachorro extends Animal {
    public String som() {
        return "Au au!";
    }
}

class Gato extends Animal {
    public String som() {
        return "Miau!";
    }
}

class Vaca extends Animal {
    public String som() {
        return "Muuu!";
    }
}

File: Lista de Exercício/Golang/Animal.go
// Questão 4 e 5
package main

import (
	"fmt"
)

type Animal interface {
	Som() string
	Tipo() string
}

type Cachorro struct{}

func (c Cachorro) Som() string {
	return "Au au!"
}

func (c Cachorro) Tipo() string {
	return "Cachorro"
}

type Gato struct{}

func (g Gato) Som() string {
	return "Miau!"
}

func (g Gato) Tipo() string {
	return "Gato"
}

// Vaca representa uma vaca
type Vaca struct{}

func (v Vaca) Som() string {
	return "Muuu!"
}

func (v Vaca) Tipo() string {
	return "Vaca"
}

func FazerBarulho(animais []Animal) {
	for _, animal := range animais {
		fmt.Printf("O %s faz: %s\n", animal.Tipo(), animal.Som())
	}
}

func Main() {
	cachorro := Cachorro{}
	gato := Gato{}
	vaca := Vaca{}

	animais := []Animal{cachorro, gato, vaca}

	FazerBarulho(animais)
}
